compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1999,6 +1999,31 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         (ty::Bool, ty::Tuple(list)) => if list.len() == 0 {
                             self.suggest_let_for_letchains(&mut err, &trace.cause, span);
                         }
+                        // Suggest providing a std::ptr::null{,_mut}() if they pass in a 0 to an *{const,mut} uint
+                        (ty::RawPtr(ptr), ty::Uint(_))  => {
+                            if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
+                                if let Ok(0) = code.parse::<u8>() {
```

Comment:
Is additional string parsing a good idea here?

We can discard all `_`, followed by then taking the subset of the str that excludes anything after an `i` or `u`, if one was present. That should handle all cases I can think of (`000_000`, `0_0usize` `0_0_iu32`, and parse already supports something like `"0000000000000"` parsing as a valid `u8`). Maybe this work has already been done somewhere?

Alternatively we can just leave that as potential future work -- after all I don't think it will be common that folks will pass in a prefixed numeric literal to an raw pointer argument.

Reply:
it's more that I don't think that parsing `span_to_snippet` is a good idea at all, see @compiler-errors' comments: https://github.com/rust-lang/rust/pull/107553#discussion_r1093963499